### PR TITLE
changes to optimise serverless execution

### PR DESCRIPTION
**Pull request by bhumilsarvaiya**
closes issue #6

### CODE REVIEW

**Comment by rishabhpoddar** [comment]
@bhumilsarvaiya please also make docs changes for this in:
- [ ] nodejs docs
- [ ] in thirdpartyemailpassword recipe docs, have a section for serverless env and in that explain how to switch this flag on and what it does and why it's needed. Then from the other recipes, have the same section, but in those, link to this section which actually has all the info

File: lib/ts/constants.ts [hunk 1]
```diff
@@ -15,3 +15,4 @@
 
 export const HEADER_RID = "rid";
 export const HEADER_FDI = "fdi-version";
+export const API_VERSION_FILE_PATH = "/tmp/apiversion.txt";
```
Comment (rishabhpoddar) [inline lib/ts/constants.ts:18]: Supposed to be `temp` not `tmp`

File: lib/ts/utils.ts [hunk 2]
```diff
@@ -205,3 +206,29 @@ export function validateTheStructureOfUserInput(
...
+        return apiVersionFromFile.toString();
+    } catch (err) {
+        return null;
+    }
+}
+
+export async function storeAPIVersionInFile(apiVersion: string) {
+    try {
+        await new Promise((resolve, reject) => {
+            writeFile(API_VERSION_FILE_PATH, apiVersion, {}, () => {
```
Comment (rishabhpoddar) [inline lib/ts/utils.ts:229]: - If the files does not exist, does it create a new one?
- Does this overwrite the file if it already exists?
- We should also check that `/temp` exists before attempting to write to it, to be sure that we do not implicitly create one.
- Abstract away reading and writing to files within `/temp` folder into it's own function that does the above check as well.

File: lib/ts/constants.ts (hunk 1 above)
Comment (rishabhpoddar) [reply lib/ts/constants.ts:18]: Change the name of the file to `supertokens-apiversion.txt` and same for the other one

File: lib/ts/querier.ts [hunk 3]
```diff
@@ -34,10 +34,16 @@ export class Querier {
     private __hosts: NormalisedURLDomain[];
     private recipe: RecipeModule | undefined;
     private rIdToCore: string | undefined;
+    private isInServerlessEnv: boolean;
```
Comment (rishabhpoddar) [inline lib/ts/querier.ts:37]: This is not being set anywhere.

File: lib/ts/utils.ts [hunk 4]
```diff
@@ -205,3 +206,29 @@ export function validateTheStructureOfUserInput(
         });
     }
 }
+
+export async function getAPIVersionFromFileIfExists(): Promise<string | null> {
```
Comment (rishabhpoddar) [inline lib/ts/utils.ts:210]: Please use `undefined` instead of `null`

File: lib/ts/querier.ts [hunk 5]
```diff
@@ -80,6 +93,9 @@ export class Querier {
             });
         }
         Querier.apiVersion = supportedVersion;
+        if (this.isInServerlessEnv) {
+            await storeAPIVersionInFile(supportedVersion);
```
Comment (rishabhpoddar) [inline lib/ts/querier.ts:97]: No need to do await here.

File: lib/ts/recipe/session/sessionRecipe.ts [hunk 6]
```diff
@@ -170,6 +182,9 @@ export default class SessionRecipe extends RecipeModule {
                 accessTokenValidity: response.accessTokenValidity,
                 refreshTokenValidity: response.refreshTokenValidity,
             };
+            if (this.checkIfInServerlessEnv()) {
+                await storeHandshakeInfoInFile(this.handshakeInfo);
```
Comment (rishabhpoddar) [inline lib/ts/recipe/session/sessionRecipe.ts:186]: Do not need to await

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/constants.ts (hunk 1 above)
Comment (bhumilsarvaiya) [reply lib/ts/constants.ts]: > Change the name of the file to supertokens-apiversion.txt and same for the other one

done

File: lib/ts/utils.ts (hunk 2 above)
Comment (bhumilsarvaiya) [reply lib/ts/utils.ts]: > If the files does not exist, does it create a new one?

No, but the logic has now been changed

> Does this overwrite the file if it already exists?

yes

> We should also check that /temp exists before attempting to write to it, to be sure that we do not implicitly create one.

yes, the directory will now be created if not already exists

> Abstract away reading and writing to files within /temp folder into it's own function that does the above check as well.

done

File: lib/ts/querier.ts (hunk 3 above)
Comment (bhumilsarvaiya) [reply lib/ts/querier.ts:38]: done

File: lib/ts/utils.ts (hunk 4 above)
Comment (bhumilsarvaiya) [reply lib/ts/utils.ts]: done

File: lib/ts/querier.ts (hunk 5 above)
Comment (bhumilsarvaiya) [reply lib/ts/querier.ts]: done

File: lib/ts/recipe/session/sessionRecipe.ts (hunk 6 above)
Comment (bhumilsarvaiya) [reply lib/ts/recipe/session/sessionRecipe.ts]: done

File: CHANGELOG.md [hunk 7]
```diff
@@ -4,6 +4,10 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [4.2.0] - 2021-03-09
+### Added
+- Optimization for serverless execution
```
Comment (rishabhpoddar) [inline CHANGELOG.md:9]: Also add "Added config to enable serverless optimisation"

File: lib/ts/constants.ts [hunk 8]
```diff
@@ -15,4 +15,4 @@
 
 export const HEADER_RID = "rid";
 export const HEADER_FDI = "fdi-version";
-export const API_VERSION_FILE_PATH = "/tmp/apiversion.txt";
+export const API_VERSION_FILE_PATH = "/tmp/supertokens-apiversion.txt";
```
Comment (rishabhpoddar) [inline lib/ts/constants.ts:18]: - Should not be `tmp`! Should be based on `process.env` (for test) or if that is undefined, should be `/temp`
- Also, change the variable name to be somehting like `SERVERLESS_CACHE_API_VERSION_FILE_PATH`

File: lib/ts/utils.ts [hunk 9]
```diff
@@ -207,28 +208,62 @@ export function validateTheStructureOfUserInput(
...
-        return apiVersionFromFile.toString();
+        return JSON.parse(dataFromFile.toString());
     } catch (err) {
-        return null;
+        return undefined;
     }
 }
 
-export async function storeAPIVersionInFile(apiVersion: string) {
+export async function storeIntoTempFile(filePath: string, data: any) {
```
Comment (rishabhpoddar) [inline lib/ts/utils.ts:227]: Change the name of the function to be `storeIntoTempFolderForServerlessCache`

File: lib/ts/utils.ts [hunk 10]
```diff
@@ -207,28 +208,62 @@ export function validateTheStructureOfUserInput(
...
+    try {
+        await createTempDirIfNotAlreadyExists();
```
Comment (rishabhpoddar) [inline lib/ts/utils.ts:229]: No.. We do not want to create the dir. It should already be there.. if it does not exist, we do nothing.

File: lib/ts/utils.ts [hunk 11]
```diff
@@ -207,28 +208,62 @@ export function validateTheStructureOfUserInput(
     }
 }
 
-export async function getAPIVersionFromFileIfExists(): Promise<string | null> {
+export async function getDataFromFileIfExists<T>(filePath: string): Promise<T | undefined> {
```
Comment (rishabhpoddar) [inline lib/ts/utils.ts:211]: - Why do you need a generic type T? Shouldn't this always return a string | undefined?
- Change the name of the function to `getDataFromFileForServerlessCache`

File: lib/ts/supertokens.ts [hunk 12]
```diff
@@ -60,6 +61,10 @@ export default class SuperTokens {
 
         this.isInServerlessEnv = config.isInServerlessEnv === undefined ? false : config.isInServerlessEnv;
 
+        if (!this.isInServerlessEnv) {
+            removeTempFiles();
```
Comment (rishabhpoddar) [inline lib/ts/supertokens.ts:65]: Write a comment here explaining why we are doing this.

File: lib/ts/utils.ts [hunk 13]
```diff
@@ -207,28 +208,62 @@ export function validateTheStructureOfUserInput(
...
+                if (err !== undefined && err !== null) {
+                    reject(err);
+                }
                 resolve(undefined);
             });
         });
     } catch (err) {}
 }
+
+export async function removeTempFiles() {
```
Comment (rishabhpoddar) [inline lib/ts/utils.ts:264]: Change the name of this function to be `removeServerlessCache`

File: lib/ts/recipe/session/sessionRecipe.ts [hunk 14]
```diff
@@ -162,8 +157,8 @@ export default class SessionRecipe extends RecipeModule {
     getHandshakeInfo = async (): Promise<HandshakeInfo> => {
         if (this.handshakeInfo === undefined) {
             if (this.checkIfInServerlessEnv()) {
-                let handshakeInfo = await getHandshakeInfoFromFileIfExists();
-                if (handshakeInfo !== null) {
+                let handshakeInfo = await getDataFromFileIfExists<HandshakeInfo>(HANDSHAKE_INFO_FILE_PATH);
```
Comment (rishabhpoddar) [inline lib/ts/recipe/session/sessionRecipe.ts:160]: Arent you supposed to use JSON.parse here?

File: lib/ts/recipe/session/sessionRecipe.ts [hunk 15]
```diff
@@ -183,7 +178,7 @@ export default class SessionRecipe extends RecipeModule {
                 refreshTokenValidity: response.refreshTokenValidity,
             };
             if (this.checkIfInServerlessEnv()) {
-                await storeHandshakeInfoInFile(this.handshakeInfo);
+                storeIntoTempFile(HANDSHAKE_INFO_FILE_PATH, this.handshakeInfo);
```
Comment (rishabhpoddar) [inline lib/ts/recipe/session/sessionRecipe.ts:181]: Arent you supposed to to do JSON.stringify here?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/session/sessionRecipe.ts (hunk 15 above)
Comment (bhumilsarvaiya) [reply lib/ts/recipe/session/sessionRecipe.ts:181]: No, done by the function storing in the file.

File: lib/ts/recipe/session/sessionRecipe.ts (hunk 14 above)
Comment (bhumilsarvaiya) [reply lib/ts/recipe/session/sessionRecipe.ts]: No, JSON parse is done by the common retrieving function

File: lib/ts/utils.ts (hunk 13 above)
Comment (bhumilsarvaiya) [reply lib/ts/utils.ts:250]: done

File: lib/ts/utils.ts (hunk 11 above)
Comment (bhumilsarvaiya) [reply lib/ts/utils.ts]: > Why do you need a generic type T? Shouldn't this always return a string | undefined

while storing apiVersion (string) or handshakeInfo (JSON), we JSON.stringify it. So when reading back from the file, when JSON.parse is used, the content of apiVersion file would result in string whereas the content of handshake info would result in JSON.

File: lib/ts/utils.ts (hunk 10 above)
Comment (bhumilsarvaiya) [reply lib/ts/utils.ts]: okay

File: lib/ts/utils.ts (hunk 9 above)
Comment (bhumilsarvaiya) [reply lib/ts/utils.ts]: done

File: lib/ts/constants.ts (hunk 8 above)
Comment (bhumilsarvaiya) [reply lib/ts/constants.ts]: okay

File: lib/ts/recipe/session/constants.ts [hunk 16]
```diff
@@ -14,4 +14,6 @@
  */
 
 export const REFRESH_API_PATH = "/session/refresh";
-export const HANDSHAKE_INFO_FILE_PATH = "/tmp/supertokens-handshakeInfo.json";
+const SERVERLESS_CACHE_BASE_FILE_PATH =
```
Comment (rishabhpoddar) [inline lib/ts/recipe/session/constants.ts:17]: Why is this being repeated here? Import from ../../constants?

File: lib/ts/recipe/session/constants.ts [hunk 17]
```diff
@@ -14,4 +14,6 @@
...
+    process.env.SERVERLESS_CACHE_BASE_FILE_PATH === undefined ? "/tmp" : process.env.SERVERLESS_CACHE_BASE_FILE_PATH;
+export const SERVERLESS_CACHE_HANDSHAKE_INFO_FILE_PATH = `${SERVERLESS_CACHE_BASE_FILE_PATH}/supertokens-handshakeInfo.json`;
```
Comment (rishabhpoddar) [inline lib/ts/recipe/session/constants.ts:19]: Remove the .json extension and also from the other file, remove the .txt extension.

File: lib/ts/utils.ts [hunk 18]
```diff
@@ -224,9 +224,8 @@ export async function getDataFromFileIfExists<T>(filePath: string): Promise<T |
     }
 }
 
-export async function storeIntoTempFile(filePath: string, data: any) {
+export async function storeIntoTempFolderForServerlessCache(filePath: string, data: any) {
     try {
-        await createTempDirIfNotAlreadyExists();
         await new Promise(async (resolve, reject) => {
```
Comment (rishabhpoddar) [inline lib/ts/utils.ts:229]: Check if the base dir exists? If it does not, do nothing..??

File: lib/ts/utils.ts (hunk 13 above)
Comment (rishabhpoddar) [reply lib/ts/utils.ts:250]: Not done.. code is still `removeTempFiles`.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/utils.ts [hunk 19]
```diff
@@ -205,3 +207,49 @@ export function validateTheStructureOfUserInput(
...
+        return JSON.parse(dataFromFile.toString());
+    } catch (err) {
+        return undefined;
+    }
+}
+
+export async function storeIntoTempFolderForServerlessCache(filePath: string, data: any) {
+    try {
+        await new Promise(async (resolve, reject) => {
+            writeFile(filePath, JSON.stringify(data), (err) => {
```
Comment (rishabhpoddar) [inline lib/ts/utils.ts:230]: You are using JSON stringify here. For the API version, you are only giving this a string, will that work?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/utils.ts (hunk 10 above)
Comment (rishabhpoddar) [reply lib/ts/utils.ts]: You are not checking if the dir exists or not.. If it does not exist, will trying to write to a file in that dir throw an error?

File: lib/ts/utils.ts (hunk 19 above)
Comment (bhumilsarvaiya) [reply lib/ts/utils.ts:230]: yes

File: lib/ts/recipe/session/constants.ts (hunk 16 above)
Comment (bhumilsarvaiya) [reply lib/ts/recipe/session/constants.ts]: okay

**Review by rishabhpoddar — APPROVED** [review]

File: lib/ts/utils.ts (hunk 10 above)
Comment (rishabhpoddar) [reply lib/ts/utils.ts]: Please reply to this @bhumilsarvaiya

Comment (bhumilsarvaiya) [reply lib/ts/utils.ts]: yes, if the dir doesn't exists, the writeFile function will throw an error

File: test/utils.js [hunk 20]
```diff
@@ -377,3 +377,17 @@ module.exports.emailVerifyTokenRequest = async function (app, accessToken, idRef
 
     return result;
 };
+
+module.exports.createTmpDirForTesting = async function () {
+    await new Promise((resolve, reject) => {
+        fs.mkdir(process.env.SERVERLESS_CACHE_BASE_FILE_PATH, (err) => {
+            if (err !== undefined && err !== null) {
+                if (err.message.includes("EEXIST")) {
+                    resolve();
```
Comment (rishabhpoddar) [inline test/utils.js:386]: shouldn't this be return resolve()?

File: test/config.test.js [hunk 21]
```diff
@@ -1248,4 +1258,98 @@ describe(`configTest: ${printPath("[test/config.test.js]")}`, function () {
         assert.strictEqual(getTopLevelDomainForSameSiteResolution("http://test.com:3567"), "test.com");
         assert.strictEqual(getTopLevelDomainForSameSiteResolution("https://test.com:3567"), "test.com");
     });
+
+    it("testing no calls to the core are happening (on supertokens.init) if the cache files exist with proper values", async function () {
+        await startST();
+        await createTmpDirForTesting(); // create dir if not already present
```
Comment (rishabhpoddar) [inline test/config.test.js:1264]: Put these two in beforeEach

File: test/config.test.js [hunk 22]
```diff
@@ -1248,4 +1258,98 @@ describe(`configTest: ${printPath("[test/config.test.js]")}`, function () {
...
+        await removeServerlessCache(); // if dir already present, it may still have some tmp files
+        {
+            let verifyHandshakeState = await ProcessState.getInstance().waitForEvent(
```
Comment (rishabhpoddar) [inline test/config.test.js:1267]: Change this test to have process state in `sendRequestHelper`

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: test/utils.js (hunk 20 above)
Comment (bhumilsarvaiya) [reply test/utils.js]: corrected

File: test/config.test.js (hunk 21 above)
Comment (bhumilsarvaiya) [reply test/config.test.js]: added

File: test/config.test.js (hunk 22 above)
Comment (bhumilsarvaiya) [reply test/config.test.js]: done

File: test/emailpassword/signinFeature.test.js [hunk 23]
```diff
@@ -39,11 +40,14 @@ let utils = require("../../lib/build/recipe/emailpassword/utils");
 const express = require("express");
 const request = require("supertest");
 const { default: NormalisedURLPath } = require("../../lib/build/normalisedURLPath");
+const { removeServerlessCache } = require("../../lib/build/utils");
 
 describe(`signinFeature: ${printPath("[test/emailpassword/signinFeature.test.js]")}`, function () {
     beforeEach(async function () {
         await killAllST();
         await setupST();
+        await createTmpDirForTesting();
```
Comment (rishabhpoddar) [inline test/emailpassword/signinFeature.test.js:49]: Please rename this function to `createServerlessCacheForTesting`

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**